src/firmware_new/boards/frankie_v1/io/proto_multicast_communication.c

```diff
@@ -0,0 +1,234 @@
+#include "firmware_new/boards/frankie_v1/io/proto_multicast_communication.h"
+
+#include <stdlib.h>
+
+#include "lwip.h"
+#include "lwip/api.h"
+#include "lwip/inet.h"
+#include "lwip/ip_addr.h"
+#include "lwip/memp.h"
+#include "lwip/opt.h"
+#include "lwip/pbuf.h"
+#include "lwip/sys.h"
+#include "lwip/tcp.h"
+#include "lwip/udp.h"
+#include "pb.h"
+#include "pb_decode.h"
+#include "pb_encode.h"
+
+// internal event to signal network interface configured
+static osEventFlagsId_t networking_event;
+
+// flag mask for network interface configured event
+// this flag is set on the networking_event to indicate "link up"
+static uint32_t NETIF_CONFIGURED = 1 << 0;
+
+// the timeout to recv a network packet
+static uint32_t NETWORK_TIMEOUT_MS = 1000;
+
+/**
+ * ProtoMulticastCommunicationProfile_t contains the information required
+ * to join a multicast group and then send or receive protobuf.
+ */
+typedef struct ProtoMulticastCommunicationProfile
+{
+    // name of the profile to pass down to other
+    // elements that may require a name (mutex, etc..)
+    const char* profile_name;
+
+    // multicast group info: the address and the port to bind to and
+    // join to send/recv proto
+    ip_addr_t multicast_address;
+    uint16_t port;
+
+    // protobuf info: the pb_field_t array fully defines the protobuf_struct,
+    // so we use a void pointer to be generic to any protobuf message.
+    void* protobuf_struct;
+    const pb_field_t* message_fields;
+    uint16_t message_max_size;
+
+    // communication_event: these events will be used to control when the networking
+    // tasks run. The networking tasks will also signal certain events.
+    osEventFlagsId_t communication_event;
+
+    // mutex to protect the protobuf struct
+    osMutexId_t profile_mutex;
+
+} ProtoMulticastCommunicationProfile_t;
+
+void io_proto_multicast_communication_init(void)
+{
+    networking_event = osEventFlagsNew(NULL);
+}
+
+ProtoMulticastCommunicationProfile_t* io_proto_multicast_communication_profile_create(
+    const char* profile_name, const char* multicast_address, uint16_t port,
+    void* protobuf_struct, const pb_field_t* message_fields, uint16_t message_max_size)
+{
+    ProtoMulticastCommunicationProfile_t* profile =
+        (ProtoMulticastCommunicationProfile_t*)malloc(
+            sizeof(ProtoMulticastCommunicationProfile_t));
+
+    const osMutexAttr_t mutex_attr = {profile_name, osMutexPrioInherit, NULL, 0U};
+
+    profile->profile_name        = profile_name;
+    profile->port                = port;
+    profile->message_fields      = message_fields;
+    profile->message_max_size    = message_max_size;
+    profile->protobuf_struct     = protobuf_struct;
+    profile->profile_mutex       = osMutexNew(&mutex_attr);
+    profile->communication_event = osEventFlagsNew(NULL);
+    ip6addr_aton(multicast_address, &profile->multicast_address);
+
+    return profile;
+}
+
+void io_proto_multicast_sender_Task(void* arg)
```

Comment:
Any chance we can use a more descriptive argument name?

Reply:
changed